### PR TITLE
Implement Phishing Detection library

### PR DESCRIPTION
**Pull request by not-a-rootkit**
Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/1204023833050360/1207138193507363/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2889
macOS PR: https://github.com/duckduckgo/macos-browser/pull/2791
What kind of version bump will this require?: Major/Minor/Patch

**Optional**:

Tech Design URL: https://app.asana.com/0/481882893211075/1207220724600204/f
CC: 

**Description**:
This PR aims to introduce the in-browser PhishingDetection library to be used by our macOS and iOS browsers:
- PhishingDetectionClient.swift
  - Handles direct HTTP communication with the BE
  - Handles converting responses into the correct data types/structs
  - Error handling from backend
- PhishingDetectionService.swift
  - Takes a client injected as dependency
  - Implements higher level data collection logic for the background tasks/schedulers
  - Stores all the data in-memory in its own instance
  - Implements the isMalicious function which contains the required logic to determine whether a site should be flagged, including:
    - URL canonicalization
    - Domain extraction
    - Domain hashing
    - Hash prefix computation
- PishingDetectionDataActivities.swift
  - Implements two objects that utilize NSBackgroundActivity to update both hashPrefix and filterSet data sets.
  - This should only expose a `run()` function that starts both tasks in the background

Outstanding questions:
 - Should the persistent storage be handled by this library or should it be passed in from the platform integration code?
   - What is the best location to store these files? Documents? Library?
 - Are there any performance enhancements we can make? 
   - What about storing hashPrefixes as a Set/Dictionary? `hashPrefix` should be unique so a Dict makes some sense.
   - Caching a certain number of API responses?

**Steps to test this PR**:
1. You'll need to use [my C-S-S branch](https://github.com/duckduckgo/content-scope-scripts/tree/tespach/phishing-error-page) to get the actual error page HTML.
2. Remember to `npm build` the C-S-S repo and update the branch in Xcode
3. You'll also need [my BSK branch](https://github.com/duckduckgo/BrowserServicesKit/tree/tespach/phishing-protection-client) to get PhishingDetectionService etc.
4. Finally, you'll need to mock the API (for now.. BE should be deployed imminently!) [Use this script.](https://dub.duckduckgo.com/gist/tespach/d1e90128083d30077edc520fa641c3ef)

Now to test:
1. Browse as normal, ensure everything loads and runs as expected
2. Now browse to https://privacy-test-pages.site - ensure the error page is not shown here.
3. Under **Security** click "An example phishing page"
4. Ensure the error page shows up
5. Now click "Advanced>Visit Site" - ensure the actual page loads
6. Refresh page, ensure the page still loads
7. Navigate backwards, ensure the Privacy Test Pages home page is shown
8. Play around with refresh/back/forward here to ensure everything works as expected
9. Open a new tab, visit http://bad.third-party.site/security/badware/phishing.html, ensure error page is thrown
10. Hit "Leave Site"
11. Ensure you're taken to the new tab page

**OS Testing**:

* [ ] iOS 14
* [ ] iOS 15
* [ ] iOS 16
* [ ] macOS 10.15
* [ ] macOS 11
* [ ] macOS 12

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

### CODE REVIEW

File: Sources/PhishingDetection/PhishingDetectionService.swift [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        self.hostname = hostname
+        self.url = url
+        self.regex = regex
+        self.hash = hash
+    }
+}
+
+public protocol PhishingDetectionServiceProtocol {
+    var filterSet: [Filter] {get set}
+    var hashPrefixes: [String] {get set}
```
Comment (SabrinaTardio) [inline Sources/PhishingDetection/PhishingDetectionService.swift:68]: Do they need to be on the protocol

File: Sources/PhishingDetection/PhishingDetectionService.swift [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    func isMalicious(url: URL) async -> Bool
+    func updateFilterSet() async
+    func updateHashPrefixes() async
+    func getMatches(hashPrefix: String) async -> [Match]
```
Comment (SabrinaTardio) [inline Sources/PhishingDetection/PhishingDetectionService.swift:72]: is this used anywhere?

File: Sources/PhishingDetection/PhishingDetectionService.swift [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    func loadData()
+    func writeData()
```
Comment (SabrinaTardio) [inline Sources/PhishingDetection/PhishingDetectionService.swift:74]: Does it need to be in the protocol?

File: Tests/PhishingDetectionTests/PhishingDetectionDataActivitiesTest.swift [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        super.setUp()
+        dataActivities = PhishingDetectionDataActivities(detectionService: mockDetectionService)
+    }
+
+    func testDataActivitySchedulerRuns() async {
+        let hpExpectation = XCTestExpectation(description: "HashPrefixDataActivity should run and update hash prefixes")
+
+        Task {
+            await dataActivities.run()
+            XCTAssertTrue(mockDetectionService.didUpdateHashPrefixes, "Hash prefixes should be updated")
```
Comment (SabrinaTardio) [inline Tests/PhishingDetectionTests/PhishingDetectionDataActivitiesTest.swift:27]: You could have both Assertion together after the await XCTWaiter().fulfillment(of: [hpExpectation], timeout: 5.0)

File: Tests/PhishingDetectionTests/PhishingDetectionDataActivitiesTest.swift [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+//
+//  Created by Thom on 04/05/2024.
+//
+
+import Foundation
+import XCTest
+@testable import BrowserServicesKit
+@testable import PhishingDetection
+
+class PhishingDetectionDataActivitiesTest: XCTestCase {
```
Comment (SabrinaTardio) [inline Tests/PhishingDetectionTests/PhishingDetectionDataActivitiesTest.swift:13]: If you can be bothered you can inject the intervals and set ito to like 1 second from the test anc check it’s called twice or something

File: Tests/PhishingDetectionTests/PhishingDetectionServiceTests.swift [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        XCTAssertEqual(service.hashPrefixes, [
+            "aa00bb11",
+            "bb00cc11",
+            "cc00dd11",
+            "dd00ee11",
+            "a379a6f6"
+        ])
+    }
+
+    func testGetMatches() async {
```
Comment (SabrinaTardio) [inline Tests/PhishingDetectionTests/PhishingDetectionServiceTests.swift:38]: This API I don’t think is needed and I don’t think it needs not be tested directly

File: Sources/PhishingDetection/PhishingDetectionService.swift (hunk 2 above)
Comment (not-a-rootkit) [reply Sources/PhishingDetection/PhishingDetectionService.swift]: Only for mocking the service when it's passed to PhishingDetectionDataActivities to test it. Might be overkill tbh we could probably remove the protocol.

**Comment by Blackmaster88** [comment]
> Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).
> 
> **Required**:
> 
> Task/Issue URL: https://app.asana.com/0/1204023833050360/1207138193507363/f iOS PR: [duckduckgo/iOS#2889](https://github.com/duckduckgo/iOS/pull/2889) macOS PR: https://github.com/duckduckgo/macos-browser/pull/2791 What kind of version bump will this require?: Major/Minor/Patch
> 
> **Optional**:
> 
> Tech Design URL: https://app.asana.com/0/481882893211075/1207220724600204/f CC:
> 
> **Description**: This PR aims to introduce the in-browser PhishingDetection library to be used by our macOS and iOS browsers:
> 
> * PhishingDetectionClient.swift
>   
>   * Handles direct HTTP communication with the BE
>   * Handles converting responses into the correct data types/structs
>   * Error handling from backend
> * PhishingDetectionService.swift
>   
>   * Takes a client injected as dependency
>   * Implements higher level data collection logic for the background tasks/schedulers
>   * Stores all the data in-memory in its own instance
>   * Implements the isMalicious function which contains the required logic to determine whether a site should be flagged, including:
>     
>     * URL canonicalization
>     * Domain extraction
>     * Domain hashing
>     * Hash prefix computation
> * PishingDetectionDataActivities.swift
>   
>   * Implements two objects that utilize NSBackgroundActivity to update both hashPrefix and filterSet data sets.
>   * This should only expose a `run()` function that starts both tasks in the background
> 
> Outstanding questions:
> 
> * Should the persistent storage be handled by this library or should it be passed in from the platform integration code?
>   
>   * What is the best location to store these files? Documents? Library?
> * Are there any performance enhancements we can make?
>   
>   * What about storing hashPrefixes as a Set/Dictionary? `hashPrefix` should be unique so a Dict makes some sense.
>   * Caching a certain number of API responses?
> 
> **Steps to test this PR**:
> 
> 1. You'll need to use [my C-S-S branch](https://github.com/duckduckgo/content-scope-scripts/tree/tespach/phishing-error-page) to get the actual error page HTML.
> 2. Remember to `npm build` the C-S-S repo and update the branch in Xcode
> 3. You'll also need [my BSK branch](https://github.com/duckduckgo/BrowserServicesKit/tree/tespach/phishing-protection-client) to get PhishingDetectionService etc.
> 4. Finally, you'll need to mock the API (for now.. BE should be deployed imminently!) [Use this script.](https://dub.duckduckgo.com/gist/tespach/d1e90128083d30077edc520fa641c3ef)
> 
> Now to test:
> 
> 1. Browse as normal, ensure everything loads and runs as expected
> 2. Now browse to https://privacy-test-pages.site - ensure the error page is not shown here.
> 3. Under **Security** click "An example phishing page"
> 4. Ensure the error page shows up
> 5. Now click "Advanced>Visit Site" - ensure the actual page loads
> 6. Refresh page, ensure the page still loads
> 7. Navigate backwards, ensure the Privacy Test Pages home page is shown
> 8. Play around with refresh/back/forward here to ensure everything works as expected
> 9. Open a new tab, visit http://bad.third-party.site/security/badware/phishing.html, ensure error page is thrown
> 10. Hit "Leave Site"
> 11. Ensure you're taken to the new tab page
> 
> **OS Testing**:
> 
> * [ ]  iOS 14
> * [ ]  iOS 15
> * [ ]  iOS 16
> * [ ]  macOS 10.15
> * [ ]  macOS 11
> * [ ]  macOS 12
> 
> ###### Internal references:
> [Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552) [Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

**Comment by Blackmaster88** [comment]
> Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).
> 
> **Required**:
> 
> Task/Issue URL: https://app.asana.com/0/1204023833050360/1207138193507363/f iOS PR: [duckduckgo/iOS#2889](https://github.com/duckduckgo/iOS/pull/2889) macOS PR: https://github.com/duckduckgo/macos-browser/pull/2791 What kind of version bump will this require?: Major/Minor/Patch
> 
> **Optional**:
> 
> Tech Design URL: https://app.asana.com/0/481882893211075/1207220724600204/f CC:
> 
> **Description**: This PR aims to introduce the in-browser PhishingDetection library to be used by our macOS and iOS browsers:
> 
> * PhishingDetectionClient.swift
>   
>   * Handles direct HTTP communication with the BE
>   * Handles converting responses into the correct data types/structs
>   * Error handling from backend
> * PhishingDetectionService.swift
>   
>   * Takes a client injected as dependency
>   * Implements higher level data collection logic for the background tasks/schedulers
>   * Stores all the data in-memory in its own instance
>   * Implements the isMalicious function which contains the required logic to determine whether a site should be flagged, including:
>     
>     * URL canonicalization
>     * Domain extraction
>     * Domain hashing
>     * Hash prefix computation
> * PishingDetectionDataActivities.swift
>   
>   * Implements two objects that utilize NSBackgroundActivity to update both hashPrefix and filterSet data sets.
>   * This should only expose a `run()` function that starts both tasks in the background
> 
> Outstanding questions:
> 
> * Should the persistent storage be handled by this library or should it be passed in from the platform integration code?
>   
>   * What is the best location to store these files? Documents? Library?
> * Are there any performance enhancements we can make?
>   
>   * What about storing hashPrefixes as a Set/Dictionary? `hashPrefix` should be unique so a Dict makes some sense.
>   * Caching a certain number of API responses?
> 
> **Steps to test this PR**:
> 
> 1. You'll need to use [my C-S-S branch](https://github.com/duckduckgo/content-scope-scripts/tree/tespach/phishing-error-page) to get the actual error page HTML.
> 2. Remember to `npm build` the C-S-S repo and update the branch in Xcode
> 3. You'll also need [my BSK branch](https://github.com/duckduckgo/BrowserServicesKit/tree/tespach/phishing-protection-client) to get PhishingDetectionService etc.
> 4. Finally, you'll need to mock the API (for now.. BE should be deployed imminently!) [Use this script.](https://dub.duckduckgo.com/gist/tespach/d1e90128083d30077edc520fa641c3ef)
> 
> Now to test:
> 
> 1. Browse as normal, ensure everything loads and runs as expected
> 2. Now browse to https://privacy-test-pages.site - ensure the error page is not shown here.
> 3. Under **Security** click "An example phishing page"
> 4. Ensure the error page shows up
> 5. Now click "Advanced>Visit Site" - ensure the actual page loads
> 6. Refresh page, ensure the page still loads
> 7. Navigate backwards, ensure the Privacy Test Pages home page is shown
> 8. Play around with refresh/back/forward here to ensure everything works as expected
> 9. Open a new tab, visit http://bad.third-party.site/security/badware/phishing.html, ensure error page is thrown
> 10. Hit "Leave Site"
> 11. Ensure you're taken to the new tab page
> 
> **OS Testing**:
> 
> * [ ]  iOS 14
> * [ ]  iOS 15
> * [ ]  iOS 16
> * [ ]  macOS 10.15
> * [ ]  macOS 11
> * [ ]  macOS 12
> 
> ###### Internal references:
> [Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552) [Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

**Pull request closed without merging**